### PR TITLE
select +/- 3 rows and cols for interpolation

### LINKED ISSUE #64: optimize image interpolation
currently dominated by the large number of "ok" pixels used for interpolating "bad" pixels in the `CloughTocher2DInterpolator`

```
     3940   86.391    0.022   87.185    0.022 interpolate.py:9(_interp_image)
       83   70.353    0.848   70.353    0.848 _lanczos.py:77(lanczos_resample_two)
      166   42.912    0.259   43.351    0.261 _wcs_utils.py:58(__call__)
        6   30.780    5.130   30.780    5.130 {method 'create_image_hdu' of '_fitsio.FITS' objects}
        8   29.599    3.700   29.719    3.715 _wcs_utils.py:53(__init__)
     1952   21.203    0.011   21.203    0.011 {method 'read' of '_io.FileIO' objects}
      179   13.035    0.073   13.040    0.073 fitslib.py:330(__init__)
       56    9.494    0.170    9.494    0.170 {method 'read_image' of '_fitsio.FITS' objects}
      199    7.891    0.040    7.891    0.040 {method 'normal' of 'mtrand.RandomState' objects}

```

### PR DESCRIPTION
**Pull request by esheldon**
closes #64 

Is a buffer of size 3 enough? 

Do we need better tests of this for a more
realistic case?

### CODE REVIEW

**Comment by esheldon** [comment]
With this PR the image interpolation is sub-dominant, but I think we might need better tests.
```
       97   82.242    0.848   82.242    0.848 _lanczos.py:77(lanczos_resample_two)
      194   47.492    0.245   48.076    0.248 _wcs_utils.py:55(__call__)
        6   29.929    4.988   29.929    4.988 {method 'create_image_hdu' of '_fitsio.FITS' objects}
        8   29.748    3.718   29.862    3.733 _wcs_utils.py:50(__init__)
     1952   18.916    0.010   18.916    0.010 {method 'read' of '_io.FileIO' objects}
      187   14.281    0.076   14.286    0.076 fitslib.py:330(__init__)
      184   12.566    0.068   16.049    0.087 interpolate.py:151(_grid_interp)
       64   11.921    0.186   11.921    0.186 {method 'read_image' of '_fitsio.FITS' objects}
      223    8.770    0.039    8.770    0.039 {method 'normal' of 'mtrand.RandomState' objects}
```

**Comment by esheldon** [comment]
OK, I had not been running those tests locally

File: pizza_cutter/slice_utils/interpolate.py [hunk 1]
```diff
@@ -74,7 +77,138 @@ def _interp_patch(*, image, bad_msk, i, j, size, buff):
...
+        pixels, shape [ngood, 2]
+    good_im:
+        the set of good image values, shape [ngood]
+    good_ind:
+        the 1d indices of the good pixels row*ncol + col
+    """
+
+    nrows, ncols = bad_msk.shape
+
+    ngood = bad_msk.size*3
```
Comment (beckermr) [inline pizza_cutter/slice_utils/interpolate.py]: This is like 3x the full image size, right? I think we only need something like `np.sum(bad_msk)*16` or something close to that.

Comment (beckermr) [reply pizza_cutter/slice_utils/interpolate.py]: Note `np.sum(bad_msk)` is the number of bad pixels.

File: pizza_cutter/slice_utils/interpolate.py [hunk 2]
```diff
@@ -74,7 +77,138 @@ def _interp_patch(*, image, bad_msk, i, j, size, buff):
...
+                if row_end > (nrows-1):
+                    row_end = nrows-1
+                if col_start < 0:
+                    col_start = 0
+                if col_end > (ncols-1):
+                    col_end = ncols-1
+
+                for rc in range(row_start, row_end+1):
+                    if rc == row:
+                        continue
```
Comment (beckermr) [inline pizza_cutter/slice_utils/interpolate.py]: Why are we skipping this?

File: pizza_cutter/slice_utils/interpolate.py [hunk 3]
```diff
@@ -74,7 +77,138 @@ def _interp_patch(*, image, bad_msk, i, j, size, buff):
...
+                    for cc in range(col_start, col_end+1):
+                        if cc == col:
```
Comment (beckermr) [inline pizza_cutter/slice_utils/interpolate.py]: Same question here. Why is this skipped?

File: pizza_cutter/slice_utils/interpolate.py [hunk 4]
```diff
@@ -74,7 +77,138 @@ def _interp_patch(*, image, bad_msk, i, j, size, buff):
...
+                            continue
+
+                        tval = bad_msk[rc, cc]
+                        if not tval:
```
Comment (beckermr) [inline pizza_cutter/slice_utils/interpolate.py]: I think we only need this test and not the row and column skips above. I think those are actually excluding pixels in the row and column of the bad pixel.

File: pizza_cutter/slice_utils/interpolate.py (hunk 1 above)
Comment (esheldon) [reply pizza_cutter/slice_utils/interpolate.py]: good point, thanks

File: pizza_cutter/slice_utils/interpolate.py (hunk 4 above)
Comment (esheldon) [reply pizza_cutter/slice_utils/interpolate.py]: yep

File: pizza_cutter/slice_utils/interpolate.py [hunk 5]
```diff
@@ -74,7 +77,128 @@ def _interp_patch(*, image, bad_msk, i, j, size, buff):
...
+    nrows, ncols = image.shape
+    npix = bad_msk.size
+
+    bad_pix, good_pix, good_im, good_ind = \
+        _get_nearby_good_pixels(image, bad_msk)
+
+    nbad = bad_pix.shape[0]
+    bm_frac = nbad/npix
+
+    if bm_frac < 0.90 and npix - nbad > 10:
```
Comment (esheldon) [inline pizza_cutter/slice_utils/interpolate.py]: @beckermr is this `npix - nbad > 10` still the right check now that we are doing the image all at once?

Comment (beckermr) [reply pizza_cutter/slice_utils/interpolate.py]: nope - probably not needed anymore

Comment (esheldon) [reply pizza_cutter/slice_utils/interpolate.py]: removed

**Pull request merged**